tests/core/engine_adapter/test_base.py

```diff
@@ -704,6 +705,314 @@ def test_merge_pandas(make_mocked_engine_adapter: t.Callable):
+            "test_updated_at": exp.DataType.Type.TIMESTAMP,
+        },
+        execution_time=datetime(2020, 1, 1, 0, 0, 0),
+    )
+
+    assert (
+        adapter.cursor.execute.call_args[0][0]
+        == parse_one(
+            """
+CREATE OR REPLACE TABLE "target" AS
+WITH "source" AS (
+  SELECT DISTINCT ON ("id")
+    "id",
+    "name",
+    "price",
+    "test_updated_at"
+  FROM (
+    SELECT
+      "id",
+      "name",
+      "price",
+      "updated_at"
+    FROM "source"
+  ) AS "raw_source"
+), "static" AS (
+  SELECT
+    "id",
+    "name",
+    "price",
+    "test_updated_at"
+  FROM "target"
+  WHERE
+    NOT "test_valid_to" IS NULL
+), "latest" AS (
+  SELECT
+    "id",
+    "name",
+    "price",
+    "test_updated_at"
+  FROM "target"
+  WHERE
+    "test_valid_to" IS NULL
+), "deleted" AS (
+  SELECT
+    "static"."id",
+    "static"."name",
+    "static"."price",
+    "static"."test_updated_at"
+  FROM "static"
+  LEFT JOIN "latest"
+    USING ("id")
+  WHERE
+    "latest"."test_valid_to" IS NULL
+), "latest_deleted" AS (
+  SELECT
+    "id",
+    MAX("test_valid_to") AS "test_valid_to"
+  FROM "deleted"
+  GROUP BY
+    "id"
+), "joined" AS (
+  SELECT
+    "latest"."id" AS "t_id",
+    "latest"."name" AS "t_name",
+    "latest"."price" AS "t_price",
+    "latest"."test_updated_at" AS "t_test_updated_at",
+    "source"."id" AS "s_id",
+    "source"."name" AS "s_name",
+    "source"."price" AS "s_price",
+    "source"."test_updated_at" AS "s_test_updated_at"
+  FROM "latest"
+  FULL JOIN "source"
+    USING ("id")
+), "non_updated_rows" AS (
+  SELECT
+    COALESCE("t_id", "s_id") AS "id",
+    COALESCE("t_name", "s_name") AS "name",
+    COALESCE("t_price", "s_price") AS "price",
+    COALESCE("t_test_updated_at", "s_test_updated_at") AS "test_updated_at",
+    CASE
+      WHEN "t_test_valid_from" IS NULL AND NOT "latest_deleted"."id" IS NULL
+      THEN CASE
+        WHEN "latest_deleted"."test_valid_to" > "s_test_updated_at"
+        THEN "latest_deleted"."test_valid_to"
+        ELSE "s_test_updated_at"
+      END
+      WHEN "t_test_valid_from" IS NULL
+      THEN CAST(\'1970-01-01 00:00:00\' AS TIMESTAMP)
+      ELSE "t_test_valid_from"
+    END AS "test_valid_from",
+    CASE
+      WHEN "s_test_updated_at" > "t_test_updated_at"
+      THEN "s_test_updated_at"
+      WHEN "s_id" IS NULL
+      THEN CAST(\'2020-01-01T00:00:00+00:00\' AS TIMESTAMP)
+      ELSE "t_test_valid_to"
+    END AS "test_valid_to"
+  FROM "joined"
+  LEFT JOIN "latest_deleted"
+    ON "joined"."s_id" = "latest_deleted"."id"
+), "updated_rows" AS (
+  SELECT
+    "s_id" AS "id",
+    "s_name" AS "name",
+    "s_price" AS "price",
+    "s_test_updated_at" AS "test_updated_at",
+    "s_test_updated_at" AS "test_valid_from",
+    NULL AS "test_valid_to"
+  FROM "joined"
+  WHERE
+    NOT "t_id" IS NULL AND NOT "s_id" IS NULL AND "s_test_updated_at" > "t_test_updated_at"
+)
+SELECT
+  *
+FROM "static"
+UNION ALL
+SELECT
+  *
+FROM "non_updated_rows"
```

Comment:
got a bit confused because non_updated_rows is getting updated if it's null and no longer null right?

Reply:
Right was just looking at it from a different perspective when I wrote it but I think the way you are framing it is more intuitive. Will update.